internal/provider/provider.go

```diff
@@ -39,37 +40,50 @@ const (
 	JujuClientID     = "client_id"
 	JujuClientSecret = "client_secret"
 	JujuCACert       = "ca_certificate"
+
+	TwoSourcesAuthWarning = "Two sources of identity for controller login"
 )
 
-// populateJujuProviderModelLive gets the controller config,
-// first from environment variables, then from a live juju
-// controller as a fallback.
-func populateJujuProviderModelLive() (jujuProviderModel, error) {
+// jujuProviderModelEnvVar gets the controller config,
+// from environment variables.
+func jujuProviderModelEnvVar() jujuProviderModel {
+	return jujuProviderModel{
+		ControllerAddrs: getEnvVar(JujuControllerEnvKey),
+		CACert:          getEnvVar(JujuCACertEnvKey),
+		ClientID:        getEnvVar(JujuClientIDEnvKey),
+		ClientSecret:    getEnvVar(JujuClientSecretEnvKey),
+		UserName:        getEnvVar(JujuUsernameEnvKey),
+		Password:        getEnvVar(JujuPasswordEnvKey),
+	}
+}
+
+func jujuProviderModelLiveDiscovery() (jujuProviderModel, bool) {
 	data := jujuProviderModel{}
 	controllerConfig, cliNotExist := juju.GetLocalControllerConfig()
-	data.ControllerAddrs = types.StringValue(getField(JujuControllerEnvKey, controllerConfig))
-	data.UserName = types.StringValue(getField(JujuUsernameEnvKey, controllerConfig))
-	data.Password = types.StringValue(getField(JujuPasswordEnvKey, controllerConfig))
-	data.ClientID = types.StringValue(getField(JujuClientIDEnvKey, controllerConfig))
-	data.ClientSecret = types.StringValue(getField(JujuClientSecretEnvKey, controllerConfig))
-	data.CACert = types.StringValue(getField(JujuCACertEnvKey, controllerConfig))
-	// Only error if a valid controller config could not be fetched
-	// from the environment variables.
-	if cliNotExist != nil && !data.valid() {
-		return data, errors.New("unable to acquire Juju controller config: no working Juju client, and environment variables are not fully set")
-	}
-
-	return data, nil
+
+	if ctrlAddrs, ok := controllerConfig[JujuControllerEnvKey]; ok && ctrlAddrs != "" {
+		data.ControllerAddrs = types.StringValue(ctrlAddrs)
+	}
+	if caCert, ok := controllerConfig[JujuCACertEnvKey]; ok && caCert != "" {
+		data.ControllerAddrs = types.StringValue(caCert)
+	}
+	if user, ok := controllerConfig[JujuUsernameEnvKey]; ok && user != "" {
+		data.ControllerAddrs = types.StringValue(user)
+	}
+	if password, ok := controllerConfig[JujuPasswordEnvKey]; ok && password != "" {
+		data.ControllerAddrs = types.StringValue(password)
+	}
```

Comment:
This seems like a copy/paste issue, because `data.ControllerAddrs` gets set in all `if`-statements.

Reply:
It is, thanks for catching it.